### ISSUE TITLE
Can't Get Return/output parameter

### ISSUE THREAD

**Issue by Honjoy** [dotnet/efcore#17538]
In ADO.NET, I can get DbDataReader and the output parameter/return parameter, if you close DbDataReader the parameter always can read it. but in EntityFrameworkCore connection I can't read parameter if have DbDataReader .

sql server:

create proc _Test_StoreProce
	@error nvarchar(max)
AS
BEGIN
	select 1 test
    set @error='some things output';
	return 123;
END

c#: 

using var cmd = dbConnection.Database.GetDbConnection().CreateCommand();
  cmd.CommandText = "_Test_StoreProce";
  cmd.CommandType = CommandType.StoredProcedure;

  var err = cmd.CreateParameter();
  err.ParameterName = "error";
  err.DbType = DbType.String;
  err.Size = 127;
  err.Direction = ParameterDirection.Output;
  cmd.Parameters.Add(err);
  var ret = cmd.CreateParameter();
  ret.ParameterName = "RC";
  ret.DbType = DbType.Int32;
  ret.Direction = ParameterDirection.ReturnValue;
  cmd.Parameters.Add(ret);

  await cmd.Connection.OpenAsync();
  using var ddr = await cmd.ExecuteReaderAsync();
  if(cmd.Parameters["RC"].Value is int s && s != 0)  //issues: cmd.Parameters["RC"].Value is null
    return new { s,m = cmd.Parameters["error"].Value }; //issues: cmd.Parameters["error"].Value is null

**Comment by ajcvickers** [comment]
@Honjoy This is just a duplicate of issue #17520. Let me try to explain again:
* The only EF code here is the first line which gets the `DbConnection`. EF isn't doing anything special here--creating the connection outside of EF would result in the same behavior.
* The rest of the code is using vanilla SqlClient/ADO.NET code. In SqlClient, parameter values are not set until after the `DbDataReader` is closed. So in this code:

```C#
using var ddr = await cmd.ExecuteReaderAsync();
if(cmd.Parameters["RC"].Value is int s && s != 0) //issues: cmd.Parameters["RC"].Value is null
return new { s,m = cmd.Parameters["error"].Value };
```
the call to `cmd.Parameters["RC"]` is happening before the data reader is closed, and hence the value is not yet set.

If the data reader is closed first, then the values are set as you expect:
```C#
using var ddr = await cmd.ExecuteReaderAsync();
ddr.Close();
if(cmd.Parameters["RC"].Value is int s && s != 0) //issues: cmd.Parameters["RC"].Value is null
return new { s,m = cmd.Parameters["error"].Value };
```
